kong/router/atc.lua

```diff
@@ -560,40 +560,49 @@ function _M:exec(ctx)
                     (sni        or "") .. (headers_key or "")
 
   local match_t = self.cache:get(cache_key)
-  if not match_t then
-    if self.cache_neg:get(cache_key) then
-      route_match_stat(ctx, "neg")
-      return nil
-    end
 
-    local req_scheme = ctx and ctx.scheme or var.scheme
+  -- pos cache hit
+  if match_t then
+    route_match_stat(ctx, "pos")
 
-    local err
-    match_t, err = self:select(req_method, req_uri, req_host, req_scheme,
-                          nil, nil, nil, nil,
-                          sni, headers)
-    if not match_t then
-      if err then
-        ngx_log(ngx_ERR, "router returned an error: ", err,
-                         ", 404 Not Found will be returned for the current request")
-      end
+    -- debug HTTP request header logic
+    add_debug_headers(var, header, match_t)
 
-      self.cache_neg:set(cache_key, true)
-      return nil
-    end
+    return match_t
```

Comment:
i'm not sure if here return early is a good idea. assume we are going to add a new behaviour for cache hit, we might forget to add in either the block from Line 565 or Line 588

Reply:
sounds reasonable. I will close this PR.